File: java_src/src/test/java/com/google/crypto/tink/signature/Ed25519PublicKeyTest.java
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.signature;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.assertThrows;

import com.google.crypto.tink.aead.ChaCha20Poly1305Key;
import com.google.crypto.tink.internal.KeyTester;
import com.google.crypto.tink.subtle.Random;
import com.google.crypto.tink.util.Bytes;
import com.google.crypto.tink.util.SecretBytes;
import java.security.GeneralSecurityException;
import org.junit.Test;
import org.junit.experimental.theories.DataPoints;
import org.junit.experimental.theories.FromDataPoints;
import org.junit.experimental.theories.Theories;
import org.junit.experimental.theories.Theory;
import org.junit.runner.RunWith;

@RunWith(Theories.class)
public final class Ed25519PublicKeyTest {

  @DataPoints("requireIdVariants")
  public static final Ed25519Parameters.Variant[] REQUIRE_ID_VARIANTS =
      new Ed25519Parameters.Variant[] {
        Ed25519Parameters.Variant.TINK,
        Ed25519Parameters.Variant.CRUNCHY,
        Ed25519Parameters.Variant.LEGACY
      };

  @Test
  public void buildNoPrefixVariantAndGetProperties() throws Exception {
    Bytes keyBytes = Bytes.copyFrom(Random.randBytes(32));

    Ed25519PublicKey key = Ed25519PublicKey.create(keyBytes);

    assertThat(key.getParameters()).isEqualTo(Ed25519Parameters.create());
    assertThat(key.getPublicKeyBytes()).isEqualTo(keyBytes);
    assertThat(key.getOutputPrefix()).isEqualTo(Bytes.copyFrom(new byte[] {}));
    assertThat(key.getIdRequirementOrNull()).isNull();
  }

  @Test
  public void buildNoPrefixVariantExplicitAndGetProperties() throws Exception {
    Bytes keyBytes = Bytes.copyFrom(Random.randBytes(32));

    Ed25519PublicKey key =
        Ed25519PublicKey.create(
            Ed25519Parameters.Variant.NO_PREFIX, keyBytes, /* idRequirement= */ null);

    assertThat(key.getParameters()).isEqualTo(Ed25519Parameters.create());
    assertThat(key.getPublicKeyBytes()).isEqualTo(keyBytes);
    assertThat(key.getOutputPrefix()).isEqualTo(Bytes.copyFrom(new byte[] {}));
    assertThat(key.getIdRequirementOrNull()).isNull();
  }

  @Test
  public void buildTinkVariantAndGetProperties() throws Exception {
    Bytes keyBytes = Bytes.copyFrom(Random.randBytes(32));

    Ed25519PublicKey key =
        Ed25519PublicKey.create(
            Ed25519Parameters.Variant.TINK, keyBytes, /* idRequirement= */ 0x0708090a);

    assertThat(key.getParameters())
        .isEqualTo(Ed25519Parameters.create(Ed25519Parameters.Variant.TINK));
    assertThat(key.getPublicKeyBytes()).isEqualTo(keyBytes);
    assertThat(key.getOutputPrefix())
        .isEqualTo(Bytes.copyFrom(new byte[] {0x01, 0x07, 0x08, 0x09, 0x0a}));
    assertThat(key.getIdRequirementOrNull()).isEqualTo(0x708090a);
  }

  @Test
  public void buildCrunchyVariantAndGetProperties() throws Exception {
    Bytes keyBytes = Bytes.copyFrom(Random.randBytes(32));

    Ed25519PublicKey key =
        Ed25519PublicKey.create(
            Ed25519Parameters.Variant.CRUNCHY, keyBytes, /* idRequirement= */ 0x0708090a);

    assertThat(key.getParameters())
        .isEqualTo(Ed25519Parameters.create(Ed25519Parameters.Variant.CRUNCHY));
    assertThat(key.getPublicKeyBytes()).isEqualTo(keyBytes);
    assertThat(key.getOutputPrefix())
        .isEqualTo(Bytes.copyFrom(new byte[] {0x00, 0x07, 0x08, 0x09, 0x0a}));
    assertThat(key.getIdRequirementOrNull()).isEqualTo(0x708090a);
  }

  @Test
  public void buildLegacyVariantAndGetProperties() throws Exception {
    Bytes keyBytes = Bytes.copyFrom(Random.randBytes(32));

    Ed25519PublicKey key =
        Ed25519PublicKey.create(
            Ed25519Parameters.Variant.LEGACY, keyBytes, /* idRequirement= */ 0x0708090a);

    assertThat(key.getParameters())
        .isEqualTo(Ed25519Parameters.create(Ed25519Parameters.Variant.LEGACY));
    assertThat(key.getPublicKeyBytes()).isEqualTo(keyBytes);
    assertThat(key.getOutputPrefix())
        .isEqualTo(Bytes.copyFrom(new byte[] {0x00, 0x07, 0x08, 0x09, 0x0a}));
    assertThat(key.getIdRequirementOrNull()).isEqualTo(0x708090a);
  }

  @Theory
  public void requireIdButIdIsNotSet_fails(
      @FromDataPoints("requireIdVariants") Ed25519Parameters.Variant variant) throws Exception {
    Bytes keyBytes = Bytes.copyFrom(Random.randBytes(32));

    assertThrows(
        GeneralSecurityException.class,
        () -> Ed25519PublicKey.create(variant, keyBytes, /* idRequirement= */ null));
  }

  @Test
  public void doesNotRequireIdButIdIsSet_fails() throws Exception {
    Bytes keyBytes = Bytes.copyFrom(Random.randBytes(32));

    assertThrows(
        GeneralSecurityException.class,
        () ->
            Ed25519PublicKey.create(
                Ed25519Parameters.Variant.NO_PREFIX, keyBytes, /* idRequirement= */ 1115));
  }

  @Test
  public void invalidKeySize() throws Exception {
    Bytes keyBytes = Bytes.copyFrom(Random.randBytes(64));

    assertThrows(GeneralSecurityException.class, () -> Ed25519PublicKey.create(keyBytes));
  }

  @Test
  public void testEqualities() throws Exception {
    Bytes keyBytes = Bytes.copyFrom(Random.randBytes(32));
    Bytes keyBytesCopy = Bytes.copyFrom(keyBytes.toByteArray());
    Bytes keyBytesDiff = Bytes.copyFrom(Random.randBytes(32));

    new KeyTester()
        .addEqualityGroup(
            "No prefix, keyBytes",
            Ed25519PublicKey.create(keyBytes),
            Ed25519PublicKey.create(
                Ed25519Parameters.Variant.NO_PREFIX, keyBytes, /* idRequirement= */ null),
            Ed25519PublicKey.create(
                Ed25519Parameters.Variant.NO_PREFIX, keyBytesCopy, /* idRequirement= */ null))
        .addEqualityGroup(
            "No prefix, different key bytes",
            Ed25519PublicKey.create(
                Ed25519Parameters.Variant.NO_PREFIX, keyBytesDiff, /* idRequirement= */ null))
        .addEqualityGroup(
            "Tink with key id 1907, keyBytes",
            Ed25519PublicKey.create(
                Ed25519Parameters.Variant.TINK, keyBytes, /* idRequirement= */ 1907),
            Ed25519PublicKey.create(
                Ed25519Parameters.Variant.TINK, keyBytesCopy, /* idRequirement= */ 1907))
        .addEqualityGroup(
            "Tink with key id 1908, keyBytes",
            Ed25519PublicKey.create(
                Ed25519Parameters.Variant.TINK, keyBytes, /* idRequirement= */ 1908))
        .addEqualityGroup(
            "Crunchy with key id 1907, keyBytes",
            Ed25519PublicKey.create(
                Ed25519Parameters.Variant.CRUNCHY, keyBytes, /* idRequirement= */ 1907))
        .addEqualityGroup(
            "Legacy with key id 1908, keyBytes",
            Ed25519PublicKey.create(
                Ed25519Parameters.Variant.LEGACY, keyBytes, /* idRequirement= */ 1907))
        .doTests();
  }

  @Test
  public void testDifferentKeyTypesEquality_fails() throws Exception {
    SecretBytes secretKeyBytes = SecretBytes.randomBytes(32);
    Bytes publicKeyBytes = Bytes.copyFrom(Random.randBytes(32));

    Ed25519PublicKey ed25519Key = Ed25519PublicKey.create(publicKeyBytes);
    ChaCha20Poly1305Key chaCha20Poly1305Key = ChaCha20Poly1305Key.create(secretKeyBytes);

    assertThat(ed25519Key.equalsKey(chaCha20Poly1305Key)).isFalse();
  }
}

File: cc/signature/ed25519_public_key_test.cc
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

#include "tink/signature/ed25519_public_key.h"

#include <string>

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/status/status.h"
#include "absl/types/optional.h"
#include "tink/partial_key_access.h"
#include "tink/signature/ed25519_parameters.h"
#include "tink/subtle/random.h"
#include "tink/util/statusor.h"
#include "tink/util/test_matchers.h"

namespace crypto {
namespace tink {
namespace {

using ::crypto::tink::test::IsOk;
using ::crypto::tink::test::StatusIs;
using ::testing::Eq;
using ::testing::TestWithParam;
using ::testing::Values;

struct TestCase {
  Ed25519Parameters::Variant variant;
  absl::optional<int> id_requirement;
  std::string output_prefix;
};

using Ed25519PublicKeyTest = TestWithParam<TestCase>;

INSTANTIATE_TEST_SUITE_P(
    Ed25519PublicKeyTestSuite, Ed25519PublicKeyTest,
    Values(TestCase{Ed25519Parameters::Variant::kTink, 0x02030400,
                    std::string("\x01\x02\x03\x04\x00", 5)},
           TestCase{Ed25519Parameters::Variant::kCrunchy, 0x01030005,
                    std::string("\x00\x01\x03\x00\x05", 5)},
           TestCase{Ed25519Parameters::Variant::kLegacy, 0x07080910,
                    std::string("\x00\x07\x08\x09\x10", 5)},
           TestCase{Ed25519Parameters::Variant::kNoPrefix, absl::nullopt, ""}));

TEST_P(Ed25519PublicKeyTest, CreateSucceeds) {
  TestCase test_case = GetParam();

  util::StatusOr<Ed25519Parameters> params =
      Ed25519Parameters::Create(test_case.variant);
  ASSERT_THAT(params, IsOk());

  std::string public_key_bytes = subtle::Random::GetRandomBytes(32);
  util::StatusOr<Ed25519PublicKey> public_key =
      Ed25519PublicKey::Create(*params, public_key_bytes,
                               test_case.id_requirement, GetPartialKeyAccess());
  ASSERT_THAT(public_key, IsOk());

  EXPECT_THAT(public_key->GetParameters(), Eq(*params));
  EXPECT_THAT(public_key->GetIdRequirement(), Eq(test_case.id_requirement));
  EXPECT_THAT(public_key->GetOutputPrefix(), Eq(test_case.output_prefix));
  EXPECT_THAT(public_key->GetPublicKeyBytes(GetPartialKeyAccess()),
              Eq(public_key_bytes));
}

TEST(Ed25519PublicKeyTest, CreateWithInvalidPublicKeyLength) {
  util::StatusOr<Ed25519Parameters> params =
      Ed25519Parameters::Create(Ed25519Parameters::Variant::kTink);
  ASSERT_THAT(params, IsOk());

  std::string public_key_bytes = subtle::Random::GetRandomBytes(31);

  EXPECT_THAT(
      Ed25519PublicKey::Create(*params, public_key_bytes,
                               /*id_requirement=*/123, GetPartialKeyAccess())
          .status(),
      StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST(Ed25519PublicKeyTest, CreateKeyWithInvalidIdRequirementFails) {
  util::StatusOr<Ed25519Parameters> no_prefix_params =
      Ed25519Parameters::Create(Ed25519Parameters::Variant::kNoPrefix);
  ASSERT_THAT(no_prefix_params, IsOk());

  util::StatusOr<Ed25519Parameters> tink_params =
      Ed25519Parameters::Create(Ed25519Parameters::Variant::kTink);
  ASSERT_THAT(tink_params, IsOk());

  std::string public_key_bytes = subtle::Random::GetRandomBytes(32);

  EXPECT_THAT(
      Ed25519PublicKey::Create(*no_prefix_params, public_key_bytes,
                               /*id_requirement=*/123, GetPartialKeyAccess())
          .status(),
      StatusIs(absl::StatusCode::kInvalidArgument));

  EXPECT_THAT(Ed25519PublicKey::Create(*tink_params, public_key_bytes,
                                       /*id_requirement=*/absl::nullopt,
                                       GetPartialKeyAccess())
                  .status(),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST_P(Ed25519PublicKeyTest, KeyEquals) {
  TestCase test_case = GetParam();

  util::StatusOr<Ed25519Parameters> params =
      Ed25519Parameters::Create(test_case.variant);
  ASSERT_THAT(params, IsOk());

  std::string public_key_bytes = subtle::Random::GetRandomBytes(32);

  util::StatusOr<Ed25519PublicKey> public_key =
      Ed25519PublicKey::Create(*params, public_key_bytes,
                               test_case.id_requirement, GetPartialKeyAccess());
  ASSERT_THAT(public_key, IsOk());

  util::StatusOr<Ed25519PublicKey> other_public_key =
      Ed25519PublicKey::Create(*params, public_key_bytes,
                               test_case.id_requirement, GetPartialKeyAccess());
  ASSERT_THAT(other_public_key, IsOk());

  EXPECT_TRUE(*public_key == *other_public_key);
  EXPECT_TRUE(*other_public_key == *public_key);
  EXPECT_FALSE(*public_key != *other_public_key);
  EXPECT_FALSE(*other_public_key != *public_key);
}

TEST(Ed25519PublicKeyTest, DifferentVariantNotEqual) {
  util::StatusOr<Ed25519Parameters> crunchy_params =
      Ed25519Parameters::Create(Ed25519Parameters::Variant::kCrunchy);
  ASSERT_THAT(crunchy_params, IsOk());

  util::StatusOr<Ed25519Parameters> tink_params =
      Ed25519Parameters::Create(Ed25519Parameters::Variant::kTink);
  ASSERT_THAT(tink_params, IsOk());

  std::string public_key_bytes = subtle::Random::GetRandomBytes(32);

  util::StatusOr<Ed25519PublicKey> public_key = Ed25519PublicKey::Create(
      *crunchy_params, public_key_bytes, /*id_requirement=*/0x01020304,
      GetPartialKeyAccess());
  ASSERT_THAT(public_key, IsOk());

  util::StatusOr<Ed25519PublicKey> other_public_key = Ed25519PublicKey::Create(
      *tink_params, public_key_bytes, /*id_requirement=*/0x01020304,
      GetPartialKeyAccess());
  ASSERT_THAT(other_public_key, IsOk());

  EXPECT_TRUE(*public_key != *other_public_key);
  EXPECT_TRUE(*other_public_key != *public_key);
  EXPECT_FALSE(*public_key == *other_public_key);
  EXPECT_FALSE(*other_public_key == *public_key);
}

TEST(Ed25519PublicKeyTest, DifferentPublicKeyBytesNotEqual) {
  util::StatusOr<Ed25519Parameters> params =
      Ed25519Parameters::Create(Ed25519Parameters::Variant::kTink);
  ASSERT_THAT(params, IsOk());

  std::string public_key_bytes1 = subtle::Random::GetRandomBytes(32);
  std::string public_key_bytes2 = subtle::Random::GetRandomBytes(32);

  util::StatusOr<Ed25519PublicKey> public_key = Ed25519PublicKey::Create(
      *params, public_key_bytes1, /*id_requirement=*/0x01020304,
      GetPartialKeyAccess());
  ASSERT_THAT(public_key, IsOk());

  util::StatusOr<Ed25519PublicKey> other_public_key = Ed25519PublicKey::Create(
      *params, public_key_bytes2, /*id_requirement=*/0x01020304,
      GetPartialKeyAccess());
  ASSERT_THAT(other_public_key, IsOk());

  EXPECT_TRUE(*public_key != *other_public_key);
  EXPECT_TRUE(*other_public_key != *public_key);
  EXPECT_FALSE(*public_key == *other_public_key);
  EXPECT_FALSE(*other_public_key == *public_key);
}

TEST(Ed25519PublicKeyTest, DifferentIdRequirementNotEqual) {
  util::StatusOr<Ed25519Parameters> params =
      Ed25519Parameters::Create(Ed25519Parameters::Variant::kTink);
  ASSERT_THAT(params, IsOk());

  std::string public_key_bytes = subtle::Random::GetRandomBytes(32);

  util::StatusOr<Ed25519PublicKey> public_key = Ed25519PublicKey::Create(
      *params, public_key_bytes, /*id_requirement=*/0x01020304,
      GetPartialKeyAccess());
  ASSERT_THAT(public_key, IsOk());

  util::StatusOr<Ed25519PublicKey> other_public_key = Ed25519PublicKey::Create(
      *params, public_key_bytes, /*id_requirement=*/0x02030405,
      GetPartialKeyAccess());
  ASSERT_THAT(other_public_key, IsOk());

  EXPECT_TRUE(*public_key != *other_public_key);
  EXPECT_TRUE(*other_public_key != *public_key);
  EXPECT_FALSE(*public_key == *other_public_key);
  EXPECT_FALSE(*other_public_key == *public_key);
}

}  // namespace
}  // namespace tink
}  // namespace crypto
